keps/sig-node/xxxx-container-oom-kill-mode/README.md

```diff
@@ -0,0 +1,729 @@
+- **Single-process or tightly-coupled applications** (e.g., databases, stateful services) require group kills to maintain consistency
+
+The current all-or-nothing approach at the node level forces administrators to choose a single behavior that may not be optimal for all processes, potentially leading to:
+
+- Unnecessary downtime for resilient multi-process applications
+- Inconsistent state in applications that expect full container restarts
+- Difficulty in migrating diverse workloads from cgroup v1 to v2 environments
+
+### Real-World Impact
+
+The transition to cgroup v2's default group kill behavior has had significant production impact, as documented in [PR #117793](https://github.com/kubernetes/kubernetes/pull/117793) and [PR #122813](https://github.com/kubernetes/kubernetes/pull/122813):
+
+- Multiple organizations reported severe impact on large containers where group kills caused unnecessary service disruptions[^1]
+- Production environments experienced issues where group kills bypassed SIGTERM/SIGKILL handlers, preventing graceful shutdown
+- Reports of potential issues affecting customers due to the behavioral change
+- Multi-process workloads experienced complete container restarts when only a single process needed to be terminated[^2]
+
+These reports led to [PR #122813](https://github.com/kubernetes/kubernetes/pull/122813), which attempted to add a kubelet flag but was closed after community discussion concluded that container-level configuration was the proper solution[^3]. The consensus was that node-level configuration cannot adequately address the needs of heterogeneous workloads.
+
+This KEP also deprecates the `singleProcessOOMKill` kubelet flag for removal in v1.38 (GA). Container-level configuration provides better granularity and eliminates the complexity of maintaining both node-level and container-level settings.
+
+### Goals
+
+- Add a per-container `oomKillMode` field to allow container-level OOM behavior configuration
+- Maintain full backward compatibility with the existing `singleProcessOOMKill` kubelet configuration during migration
+- Deprecate and remove `singleProcessOOMKill` kubelet flag by v1.38 (GA) to simplify configuration
+
+### Non-Goals
+
+- Provide cgroups v1 and Windows support
+
+## Proposal
+
+Add an `oomKillMode` field to the Container specification in the core v1 API. This design choice is deliberate and based on several key considerations:
+
+### Why Container-level Configuration?
+
+1. **Granular Control**: Different containers in the same pod may have different OOM requirements:
+
+   - A sidecar container might benefit from single-process kills
+   - The main application container might require group kills for consistency
+
+2. **Natural Alignment**: OOM behavior is inherently a container-level concern:
+
+   - Memory limits are set per container
```

Comment:
I think the pod-level resource feature also introduces memory limit at pod level. What's the default cgroupv2 behavior today if pod-level memory limit is reached? Do all containers and their processes within the pod get killed?

cc @ndixita